beacon-chain/rpc/service.go

```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
 	return &pb.IndexResponse{Index: index}, nil
 }
 
-// LatestAttestation streams the latest processed attestations to the rpc clients.
-func (s *Service) LatestAttestation(req *empty.Empty, stream pb.BeaconService_LatestAttestationServer) error {
-	sub := s.attestationService.IncomingAttestationFeed().Subscribe(s.incomingAttestation)
-	defer sub.Unsubscribe()
+// ValidatorAssignment streams validator assignments every slot to clients that request
+// to watch a subset of public keys in the CrystallizedState's active validator set.
+func (s *Service) ValidatorAssignment(req *pb.ValidatorAssignmentRequest, stream pb.ValidatorService_ValidatorAssignmentServer) error {
+	// If the genesis time was at 12:00:00PM a if the SlotDuration is set to
+	// 8 seconds. We can accomplish this by using utils.WaitUntilTimestamp
+	// and passing in the desired slot duration as an argument.
+	//
+	// Instead of utilizing params.SlotDuration, we utilize a property of
+	// RPC service struct called slotAlignmentDuration. This value
+	// can be set to 0 seconds as a parameter in tests.
+	// Otherwise, tests would sleep.
+	utils.WaitUntilTimestamp(s.slotAlignmentDuration)
+
+	// Right after the timestamps are aligned, we start a new ticker
+	// that fires exactly every params.SlotDuration*time.Second. This ensures
+	// a consistent "heartbeat" to keep track of every beacon slot.
+	tickerChan := time.NewTicker(s.slotAlignmentDuration).C
```

Comment:
Do we want any lead time? 

Shouldn't a validator know that they are assigned to a slot some amount of time prior to the slot so they can connect to peers on the assigned shard?

Reply:
This can be handled on the validator side. The validator will constantly listen for updates every slot from this stream and can then determine its role would happen N slots from the current one through that.